app/models/workers_scaling/algorithm_runner.rb

```diff
@@ -3,96 +3,69 @@ module WorkersScaling
-    # Otherwise returns nil
-    def self.get(experiment_id)
-      LOGGER.debug "Getting runner from cache, experiment id: #{experiment_id}"
-      return @@cache[experiment_id.to_s] if @@cache
-    end
+    RUNNER_INTERVAL = 15.seconds
+    THREADS_NUMBER = 4
 
     ##
-    # Registers instance of AlgorithmRunner for experiment_id
-    def self.put(experiment_id, runner)
-      @@cache[experiment_id.to_s] = runner
-      LOGGER.debug "Adding runner to cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Unregisters instance of AlgorithmRunner for experiment_id
-    def self.delete(experiment_id)
-      @@cache.delete(experiment_id.to_s) if @@cache
-      LOGGER.debug "Deleting runner from cache, experiment id: #{experiment_id}"
-    end
-
-    ##
-    # Expects:
-    # * instance of Experiment
-    # * instance of Algorithm for given Experiment
-    # * interval between decision loop executions in seconds
-    def initialize(experiment, algorithm, interval)
-      @experiment = experiment
-      @algorithm = algorithm
-      @interval = interval
-      @mutex = Mutex.new
-
-      self.class.put(@experiment.id, self)
-    end
-
-    ##
-    # Initializes Algorithm, then starts its decision loop
-    # Stops when there is no next execution time set
-    def start
+    # Runs periodically and loads some data from all saved Algorithms
+    # If next_execution_time has passed, starts execute_and_schedule for this algorithm
+    def self.start
       Thread.new do
         begin
-          @algorithm.initial_deployment
-          @next_execution_time = Time.now + @interval
-
-          @mutex.synchronize do
-            catch :finished do
-              until @next_execution_time.nil?
+          while true do
+            LOGGER.debug 'Entering Algorithm Runner loop'
+            work_queue = Queue.new
+            Algorithm.get_all_algorithms_times.each do |experiment_id, next_execution_time|
+              if next_execution_time <= Time.now
+                work_queue.push(experiment_id)
+              end
+            end
 
-                while @next_execution_time > Time.now do
-                  @mutex.sleep @next_execution_time - Time.now
-                  throw :finished if @next_execution_time.nil?
+            threads = (1..THREADS_NUMBER).map do
+              Thread.new do
+                until work_queue.empty?
+                  execute_and_schedule(work_queue.pop)
                 end
-
-                execute_and_schedule
               end
             end
-          end
 
-          self.class.delete(@experiment.id)
+            threads.each &:join
+
+            sleep(RUNNER_INTERVAL)
+          end
         rescue => e
-          LOGGER.error "Exception occurred during workers scaling algorithm: #{e.to_s}\n#{e.backtrace.join("\n")}"
+          LOGGER.error "Exception occurred during Algorithm Runner loop: #{e.to_s}\n#{e.backtrace.join("\n")}"
```

Comment:
Runner umiera tylko wtedy, kiedy coś się stanie w jego wątku. Może umrzeć tylko kiedy:
* zepsuje się wbudowana kolejka - mało prawdopodobne
* coś się podzieje przy startowaniu wątków - mało prawdopodobne, tylko przy uruchamianiu systemu
* zepsuje się baza - wtedy nie ma co go podnosić
* w bazie pojawi się rekord, który nie ma experiment_id albo next_execution_time - trzeba by w kodzie każde wyjęcie z osobna obłożyć rescue, a i tak to się wydarzy tylko w przypadku wpisania rekordu ręcznie przez konsolę

Jeżeli zamiast tego wybuchnie initial deployment - sytuacja jakby nie stworzono algorytmu.

Jeżeli wybuchnie experiment_status_check - umrze jeden wątek. Tutaj można dodać ratowanie go, żeby nie umierał - chcemy takie coś? Wtedy sytuacja będzie następująca: mamy n dobrych algorytmów i jeden rzucający wyjątkiem. Za każdym razem, kiedy ma być wykonany ten zły jest on wyciągany z kolejki. Następnie wątek go uruchamia, łapie wyjątek i wraca do normalnej pracy. Populacja wątków zostaje bez zmian, niestety ten algorytm będzie powodował wyjątki za każdym razem, więc co przejście będzie trafiał do kolejki i wyrzucał na chwilę jeden wątek.